### PR TITLE
refactor: enable TS strict mode

### PR DESCRIPTION
**Pull request by kanadgupta**
## 🧰 Changes

Enables the `strict` typescript flag in this repository. Any specific callouts will be flagged in [the comments below](https://github.com/readmeio/rdme/pull/889#discussion_r1326439386).

## 🧬 QA & Testing

Doing this exercise likely added type safety and error handling around weird API edge cases but I'd rather write the tests and do the QA if these edge cases become a problem. There were basically no changes to our existing tests so if those pass then we should be in good shape.

### CODE REVIEW

File: src/cmds/categories/create.ts
Comment (kanadgupta) [inline src/cmds/categories/create.ts:1]: the diff on this file is a bit wild but there aren't any actual changes to the code — just some refactoring to get rid of the unnecessary nested functions because TS doesn't like those (and neither do i lol)

File: src/cmds/openapi/index.ts [hunk 1]
```diff
@@ -222,11 +223,8 @@ export default class OpenAPICommand extends Command {
     const options: RequestInit = {
       headers: cleanHeaders(
```
Comment (kanadgupta) [inline src/cmds/openapi/index.ts:224]: I refactored the signature of our `cleanHeaders` function so the second parameter is a `version` which we use to construct our `x-readme-version` request header. You'll see a lot of refactors related to that in this PR.

File: src/lib/baseCommand.ts [hunk 2]
```diff
@@ -11,11 +10,16 @@ import isCI from './isCI';
...
+
+export type AuthenticatedCommandOptions<T> = Omit<ZeroAuthCommandOptions<T>, 'key'> & {
+  key: string;
   version?: string;
-} & CommandLineOptions;
+};
+
+export type ZeroAuthCommandOptions<T> = T & {
+  github?: boolean;
+} & { key?: never };
```
Comment (kanadgupta) [inline src/lib/baseCommand.ts:22]: split out our command options for commands that do and do not require auth so we could assert that commands with zero auth do **not** accept a key parameter. TS is cool sometimes

File: src/lib/baseCommand.ts [hunk 3]
```diff
@@ -11,11 +10,16 @@ import isCI from './isCI';
...
-  key?: string;
+export type CommandOptions<T> = ZeroAuthCommandOptions<T> | AuthenticatedCommandOptions<T>;
+
+export type AuthenticatedCommandOptions<T> = Omit<ZeroAuthCommandOptions<T>, 'key'> & {
+  key: string;
   version?: string;
-} & CommandLineOptions;
+};
+
+export type ZeroAuthCommandOptions<T> = T & {
```
Comment (erunion) [inline src/lib/baseCommand.ts:20]: Since we pass in `<{}>` to this generic in places where we don't have any options do you want to default `T` to that?

```suggestion
export type ZeroAuthCommandOptions<T = {}> = T & {
```

File: src/lib/baseCommand.ts [hunk 4]
```diff
@@ -11,11 +10,16 @@ import isCI from './isCI';
 import { debug, info, warn } from './logger';
 import loginFlow from './loginFlow';
 
-export type CommandOptions<T> = T & {
-  github?: boolean;
-  key?: string;
+export type CommandOptions<T> = ZeroAuthCommandOptions<T> | AuthenticatedCommandOptions<T>;
```
Comment (erunion) [inline src/lib/baseCommand.ts:13]:
```suggestion
export type CommandOptions<T = {}> = ZeroAuthCommandOptions<T> | AuthenticatedCommandOptions<T>;
```

File: src/lib/baseCommand.ts [hunk 5]
```diff
@@ -11,11 +10,16 @@ import isCI from './isCI';
...
+
+export type AuthenticatedCommandOptions<T> = Omit<ZeroAuthCommandOptions<T>, 'key'> & {
```
Comment (erunion) [inline src/lib/baseCommand.ts:15]:
```suggestion
export type AuthenticatedCommandOptions<T = {}> = Omit<ZeroAuthCommandOptions<T>, 'key'> & {
```

File: src/cmds/logout.ts [hunk 6]
```diff
@@ -16,7 +16,7 @@ export default class LogoutCommand extends Command {
     this.args = [];
   }
 
-  async run(opts: CommandOptions<{}>) {
+  async run(opts: ZeroAuthCommandOptions<{}>) {
```
Comment (erunion) [inline src/cmds/logout.ts:19]:
```suggestion
  async run(opts: ZeroAuthCommandOptions) {
```

File: src/cmds/oas.ts [hunk 7]
```diff
@@ -16,7 +16,7 @@ export default class OASCommand extends Command {
     this.args = [];
   }
 
-  async run(opts: CommandOptions<{}>) {
+  async run(opts: ZeroAuthCommandOptions<{}>) {
```
Comment (erunion) [inline src/cmds/oas.ts:19]:
```suggestion
  async run(opts: ZeroAuthCommandOptions) {
```

File: src/cmds/versions/delete.ts [hunk 8]
```diff
@@ -24,7 +24,7 @@ export default class DeleteVersionCommand extends Command {
     ];
   }
 
-  async run(opts: CommandOptions<{}>) {
+  async run(opts: AuthenticatedCommandOptions<{}>) {
```
Comment (erunion) [inline src/cmds/versions/delete.ts:27]:
```suggestion
  async run(opts: AuthenticatedCommandOptions) {
```

File: src/cmds/versions/index.ts [hunk 9]
```diff
@@ -33,7 +33,7 @@ export default class VersionsCommand extends Command {
     ];
   }
 
-  async run(opts: CommandOptions<{}>) {
+  async run(opts: AuthenticatedCommandOptions<{}>) {
```
Comment (erunion) [inline src/cmds/versions/index.ts:36]:
```suggestion
  async run(opts: AuthenticatedCommandOptions) {
```

File: src/cmds/whoami.ts [hunk 10]
```diff
@@ -18,7 +18,7 @@ export default class WhoAmICommand extends Command {
     this.args = [];
   }
 
-  async run(opts: CommandOptions<{}>) {
+  async run(opts: ZeroAuthCommandOptions<{}>) {
```
Comment (erunion) [inline src/cmds/whoami.ts:21]:
```suggestion
  async run(opts: ZeroAuthCommandOptions) {
```

File: src/index.ts [hunk 11]
```diff
@@ -134,9 +135,9 @@ export default function rdme(rawProcessArgv: NodeJS.Process['argv']) {
 
     cmdArgv = { key, ...cmdArgv };
 
-    return bin.run(cmdArgv).then((msg: string) => {
+    return bin.run(cmdArgv as CommandOptions<{}>).then((msg: string) => {
```
Comment (erunion) [inline src/index.ts:138]:
```suggestion
    return bin.run(cmdArgv as CommandOptions).then((msg: string) => {
```

File: src/lib/createGHA/index.ts [hunk 12]
```diff
@@ -61,11 +61,13 @@ function getKey(args: OptionDefinition[], opts: CommandOptions<{}>): string | fa
 function constructCmdString(
   command: keyof typeof commands,
   args: OptionDefinition[],
-  opts: CommandOptions<Record<string, string | boolean | undefined>>,
+  opts: CommandOptions<{}>,
```
Comment (erunion) [inline src/lib/createGHA/index.ts:64]:
```suggestion
  opts: CommandOptions,
```

File: src/lib/prompts.ts [hunk 13]
```diff
@@ -13,6 +13,8 @@ interface Spec {
   title: string;
 }
 
+export type OpenapiPromptOptions = 'create' | 'update';
```
Comment (erunion) [inline src/lib/prompts.ts:16]: super nit but can you rename this?

```suggestion
export type OpenAPIPromptOptions = 'create' | 'update';
```

**Review by erunion — APPROVED** [review]

File: src/lib/prompts.ts (hunk 13 above)
Comment (kanadgupta) [reply src/lib/prompts.ts]: i always go back and forth on these things lol
